internal/cmd/shell.go

```diff
@@ -90,7 +91,113 @@ func shell(cmd *cobra.Command, cmdArgs []string) error {
+		} else {
+			// Print out the reasons the rebuild is required. Newlines don't work well with
+			// the prompt label, so printing it above the prompt will have to do.
+			clog.LogWarn("rebuild required", requiredRebuilds...)
+			rebuildReason = fmt.Sprintf("Environment %q requires a rebuild to work correctly.", env.Name)
+		}
+
+		confirm := promptui.Prompt{
+			Label:     rebuildReason + " Do you wish to rebuild it now? (this will take a moment)",
+			IsConfirm: true,
+		}
+		if _, err := confirm.Run(); err != nil {
+			// User selected not to rebuild :(
+			return clog.Fatal(
+				"environment is not ready for use",
+				"environment requires a rebuild",
+				fmt.Sprintf("its current status is %q", env.LatestStat.ContainerStatus),
+				clog.BlankLine,
+				clog.Tipf("run \"coder envs rebuild %s --follow\" to start the environment", env.Name),
+			)
+		}
+
+		// Start the rebuild
+		if err = client.RebuildEnvironment(ctx, env.ID); err != nil {
+			return err
+		}
+
+		fallthrough // Fallthrough to watching the logs
+	case env.LatestStat.ContainerStatus == coder.EnvironmentCreating:
+		// Environment is in the process of being created, just trail the logs
+		// and wait until it is done
+		clog.LogInfo(fmt.Sprintf("Environment %q is being rebuilt to be ready for use.", env.Name))
+
+		// Watch the rebuild.
+		// TODO: (@emyrk) Does trailBuildLogs ensure the container is "ON" when it returns?
+		//		Or do I need to poll `findEnv` until it is 'ON'
+		if err = trailBuildLogs(ctx, client, env.ID); err != nil {
+			return err
+		}
+		// newline after trailBuildLogs to place user on a fresh line for their shell
+		fmt.Println()
+
+		// Find the status of the environment again for use outside the switch statement.
+		env, err = findEnv(ctx, client, envName, coder.Me)
+		if err != nil {
+			return err
+		}
+	case env.LatestStat.ContainerStatus == coder.EnvironmentFailed:
+		// A failed container might just keep re-failing. I think it should be investigated by the user
+		return clog.Fatal("the environment has failed to come online",
+			fmt.Sprintf("environment %q is not running", env.Name),
+			fmt.Sprintf("its current status is %q", env.LatestStat.ContainerStatus),
+
+			clog.BlankLine,
+			clog.Tipf("take a look at the build logs to determine what went wrong"),
+			clog.Tipf("run \"coder envs rebuild %s --follow\" to attempt to rebuild the environment", env.Name),
+		)
+	}
+
+	// TODO: @emyrk I can defer to `runCommand` to check this condition.
+	//		The error message is odd that I try to rebuild, and when I fail,
+	//		it tells the user to just try again.
+	if env.LatestStat.ContainerStatus != coder.EnvironmentOn {
+		return clog.Fatal(
+			"the environment is currently offline",
+			fmt.Sprintf("environment %q is not running", env.Name),
+			fmt.Sprintf("its current status is %q", env.LatestStat.ContainerStatus),
+			clog.BlankLine,
+			clog.Tipf("run \"coder envs rebuild %s --follow\" to start the environment", env.Name),
+		)
+	}
+
```

Comment:
this function definitely needs a reduction in complexity.

Reply:
Factored out a lot of it. I was in the midst of factoring. Launched that draft PR a bit too quick